Add namespaces to nf-lang (#6176)
nextflow-io/nextflow: docs/channel.md

# Channels

> In Nextflow, **channels** are the key data structure used to facilitate the dataflow dependencies between each step (i.e. {ref}`process <process-page>`) in a pipeline.

Reviewer: ```suggestion
In Nextflow, **channels** are the key data structures that facilitate the dataflow dependencies between each step (i.e. {ref}`process <process-page>`) in a pipeline.
```

There are two kinds of channels, *queue channels* and *value channels*. Channels are created using *channel factories* and transformed using *channel operators*.

(channel-type-queue)=

## Queue channels

A *queue channel* is a channel that *emits* an asynchronous sequence of values.

> A queue channel can be created by channel factories ({ref}`channel.of <channel-of>`, {ref}`channel.fromPath <channel-path>`, etc), operators ({ref}`operator-map`, {ref}`operator-filter`, etc), and processes (see {ref}`Process outputs <process-output>`).

Reviewer: ```suggestion
A queue channel can be created by channel factories (e.g., {ref}`channel.of <channel-of>` and {ref}`channel.fromPath <channel-path>`), operators (e.g., {ref}`operator-map` and  {ref}`operator-filter`), and processes (see {ref}`Process outputs <process-output>`).
```

Reviewer: Ect is unclear because it's not established which channel factories can create queue channels.

Author: Agreed, thanks

The values in a queue channel cannot be accessed directly -- they can only be accessed by passing the channel as input to an operator or process. For example:

```nextflow
channel.of(1, 2, 3).view { v -> "queue channel emits ${v}" }
```

```console
queue channel emits 1
queue channel emits 2
queue channel emits 3
```

(channel-type-value)=

## Value channels

A *value channel* is a channel that is *bound* to an asynchronous value.

A value channel can be created with the {ref}`channel.value <channel-value>` factory, certain operators

> ({ref}`operator-collect`, {ref}`operator-reduce`, etc), and processes (under {ref}`certain conditions <process-out-singleton>`).

Reviewer: ```suggestion
(e.g., {ref}`operator-collect` and {ref}`operator-reduce`), and processes (under {ref}`certain conditions <process-out-singleton>`).
```
